plugins/auth-backend/src/identity/MountedTokenIssuer.ts

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AnyJWK, TokenIssuer, TokenParams } from './types';
+import { promises as fs } from 'fs';
+import {
+  JWK,
+  importSPKI,
+  importPKCS8,
+  exportJWK,
+  SignJWT,
+  importJWK,
+} from 'jose';
+import { parseEntityRef } from '@backstage/catalog-model';
+import { AuthenticationError } from '@backstage/errors';
+import { KeyLike } from 'jose/dist/types/types';
+import { LoggerService } from '@backstage/backend-plugin-api';
+
+const MS_IN_S = 1000;
+const DEFAULT_ALGORITHM = 'ES256';
+const DEFAULT_KEY_ID = '1';
+
+export type Config = {
+  publicKeyFile: string;
+  privateKeyFile: string;
+  keyId?: string;
+  algorithm?: string;
+};
+
+export type Options = {
+  logger: LoggerService;
+  /** Value of the issuer claim in issued tokens */
+  issuer: string;
+  /** Path to the public key file in the SPKI format */
+  publicKeyFile: string;
+  /** Path to the matching private key in the PKCS#8 format */
+  privateKeyFile: string;
+  /** Expiration time of the JWT in seconds */
+  sessionExpirationSeconds: number;
+  /** id to uniquely identify this key within the JWK set, defaults to '1' */
+  keyId?: string;
+  /** JWS "alg" (Algorithm) Header Parameter value. Defaults to ES256.
+   * Must match the algorithm used to generate the keys in the provided files
+   */
+  algorithm?: string;
+};
+
+/**
+ * A token issuer that issues tokens from a predefined
+ * public/private key pair found at the given paths.
+ * The private key should be represented using the PKCS#8 format,
+ * while the public key should be in the SPKI format.
+ *
+ * You can generate a public and private key pair, using
+ * openssl:
+ *
+ * Generate a private key using the ES256 algorithm
+ * ```sh
+ * openssl ecparam -name prime256v1 -genkey -out private.ec.key
+ * ```
+ * Convert it to PKCS#8 format
+ * ```sh
+ * openssl pkcs8 -topk8 -inform PEM -outform PEM -nocrypt -in private.ec.key -out private.key
+ * ```
+ * Extract the public key
+ * ```sh
+ * openssl ec -inform PEM -outform PEM -pubout -in private.key -out public.key
+ * ```
+ *
+ * Provide the paths to private.key and public.key as the respective
+ * private and public key paths in the MountedTokenIssuer.create(...) method.
+ */
+export class MountedTokenIssuer implements TokenIssuer {
+  private readonly issuer: string;
+  private readonly logger: LoggerService;
+  private readonly publicKey: JWK;
+  private readonly privateKey: JWK;
+  private readonly sessionExpirationSeconds: number;
+
+  public static async create(options: Options): Promise<MountedTokenIssuer> {
+    const algorithm = options.algorithm ?? DEFAULT_ALGORITHM;
+    const keyId = options.keyId ?? DEFAULT_KEY_ID;
+    const publicKey = await this.loadPublicKeyFromFile(
+      options.publicKeyFile,
+      keyId,
+      algorithm,
+    );
+    const privateKey = await this.loadPrivateKeyFromFile(
```

Comment:
Is there an argument to be made, that these should be re-read every now and then? At least upon actively detecting that they changed? Thinking about whether it's a thing to perform rotation of already-mounted keys

Reply:
I think changing a key at runtime is problematic. Not sure if this is true in general. But in our infrastructure, first new sessions would 403 since Istio doesn't know the new key yet. Then when Istio updates its cache the old sessions would start 403-ing since Istio forgot about the old key.

This could be solved by remembering the old key for some time and wait a while before using the new key to sign keys. But that is a mechanism that I would hope to avoid since its hard to tell when its safe to forget or use a new key. 

From best practices I could find these keys can safely live for a year or more. Key rotation could be like scheduled maintenance where the keys are changed and the pods restarted outside of office hours.


What do you think holds true in the more general sense, outside our specific infrastructure?

Reply:
Hey @freben I believe (hope) this is the last conversation that this MR is waiting. What would help moving this forward. :D